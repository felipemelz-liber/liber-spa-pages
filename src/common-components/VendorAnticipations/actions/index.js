import vendorAdvances from './VendorAdvancesActions';

export default {
  vendorAdvances,
};
