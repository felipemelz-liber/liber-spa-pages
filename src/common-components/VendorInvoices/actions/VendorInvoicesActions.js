import axios from 'axios';
import {
  INVOICES_URL,
  INVOICES_ACTION_CABLE_CHANNEL,
  INVOICES_ACTION_CABLE_ROOM,
  INVOICES_REQUEST_DOWNLOAD_URL,
  INVOICES_REQUEST_ANTICIPATION_URL,
  ADMIN_INVOICES_URL,
  INVOICES_QUOTE_URL,
  INVOICES_QUOTE_ALL_URL,
} from '../urls';
import { convertToCamelCase, dispatchDataLayerEvent } from '../../../vendor/Utils';
import { addRateByPeriodPercentage, getSumFaceValue, debounce } from '../utils';

export const STATUS_FILTER = 'q[status_eq]';
export const STATUS_IN_FILTER = 'q[status_in]';
export const NUMBER_FILTER = 'q[number_in]';
export const CNPJ_FILTER = 'q[emitter_cnpj_eq]';
export const EXPIRES_EQ_FILTER = 'q[expires_at_eq]';
export const EXPIRES_GT_FILTER = 'q[expires_at_gteq]';
export const EXPIRES_LT_FILTER = 'q[expires_at_lteq]';

export const AVAILABLE = 2;
export const BLOCKED = 3;

export const FETCH_INVOICES = 'FETCH_INVOICES';
export const FETCH_INVOICES_FAILED = 'FETCH_INVOICES_FAILED';
export const ANTICIPATE_ALL_INVOICES = 'ANTICIPATE_ALL_INVOICES';
export const ANTICIPATE_INVOICES = 'ANTICIPATE_INVOICES';
export const ANTICIPATION_SUCCESS = 'ANTICIPATION_SUCCESS';
export const ANTICIPATION_ERROR = 'ANTICIPATION_ERROR';
export const UPDATE_INVOICES = 'UPDATE_INVOICES';
export const UPDATE_FILTERS = 'UPDATE_FILTERS';
export const TOGGLE_PAGE_SELECTION = 'TOGGLE_PAGE_SELECTION';
export const TOGGLE_INVOICE_SELECTION = 'TOGGLE_INVOICE_SELECTION';
export const CHANGE_PAGE = 'CHANGE_PAGE';
export const CHANGE_LISTING = 'CHANGE_LISTING';
export const QUOTE_ALL_INVOICES = 'QUOTE_ALL_INVOICES';
export const INVOICE_SIMULATION_QUOTE_ALL_SUCCESS = 'INVOICE_SIMULATION_QUOTE_ALL_SUCCESS';
export const CLEAR_INVOICES = 'CLEAR_INVOICES';
export const CHANGE_FILTERS = 'CHANGE_FILTERS';
export const QUOTE_SUCCESS = 'QUOTE_SUCCESS';
export const QUOTE_INVOICES = 'QUOTE_INVOICES';
export const REQUEST_SIMULATION_REPORT = 'REQUEST_SIMULATION_REPORT';
export const INVOICE_SIMULATION_SUBSCRIBE = 'INVOICE_SIMULATION_SUBSCRIBE';
export const INVOICE_REPORT_SIMULATION_SUBSCRIBE_RECEIVED =
  'INVOICE_REPORT_SIMULATION_SUBSCRIBE_RECEIVED';
export const VENDOR_INVOICES_UPDATE_SORT = 'VENDOR_INVOICES_UPDATE_SORT';
export const UPDATE_VENDOR_AVAILABLE_LIMIT = 'UPDATE_VENDOR_AVAILABLE_LIMIT';
export const UPDATE_SHOW_LIMIT_ERROR = 'UPDATE_SHOW_LIMIT_ERROR';
export const INSUFFICIENT_ANTICIPATION_LIMIT = 'INSUFFICIENT_ANTICIPATION_LIMIT';
export const LIMITS_SUBSCRIBE = 'LIMITS_SUBSCRIBE';
export const UPDATE_QUOTE_FACE_VALUE = 'UPDATE_QUOTE_FACE_VALUE';
export const UPDATE_SIMULATION_TIMEOUT_ID = 'UPDATE_SIMULATION_TIMEOUT_ID';

const getInvoicesUrl = (isAdmin, vendorId) =>
  isAdmin ? ADMIN_INVOICES_URL.replace(':id', vendorId) : INVOICES_URL;

export const subscribeInvoices = () => (dispatch, getState) => {
  const {
    vendorInvoices: { vendorId },
  } = getState();
  dispatch({
    type: INVOICE_SIMULATION_SUBSCRIBE,
    channel: INVOICES_ACTION_CABLE_CHANNEL,
    room: INVOICES_ACTION_CABLE_ROOM + vendorId,
  });
};

export const subscribeToLimits = () => (dispatch, getState) => {
  const {
    vendorInvoices: { roomId },
  } = getState();
  dispatch({
    type: LIMITS_SUBSCRIBE,
    channel: 'AnticipationLimitsChannel',
    roomId,
  });
};

export const updateInvoices = data => {
  const { invoices, pagination, rateByPeriod } = data;
  let formattedInvoices = invoices.map(convertToCamelCase);

  if (rateByPeriod) {
    formattedInvoices = formattedInvoices.map(addRateByPeriodPercentage);
  }

  return {
    type: UPDATE_INVOICES,
    payload: {
      invoices: formattedInvoices,
      pagination,
      rateByPeriod,
    },
  };
};

export const fetchInvoices = () => (dispatch, getState) => {
  dispatch({ type: FETCH_INVOICES });
  const {
    vendorInvoices: {
      filters,
      pagination: { current: page, per },
      token,
      isAdmin,
      vendorId,
      sorting,
    },
  } = getState();

  const filterByAvailable = isAdmin
    ? { [STATUS_IN_FILTER]: [AVAILABLE, BLOCKED] }
    : { [STATUS_FILTER]: AVAILABLE };

  return axios
    .get(getInvoicesUrl(isAdmin, vendorId), {
      headers: { Authorization: `Bearer ${token}` },
      params: { ...sorting, ...filters, ...filterByAvailable, page, per },
    })
    .then(payload => {
      const {
        data: { data },
      } = payload;

      dispatch(updateInvoices(data));
    })
    .catch(() => dispatch({ type: FETCH_INVOICES_FAILED }));
};

export const togglePageSelection = () => ({ type: TOGGLE_PAGE_SELECTION });

export const toggleInvoiceSelection = invoice => ({
  type: TOGGLE_INVOICE_SELECTION,
  payload: { invoice },
});

export const setPage = page => ({ type: CHANGE_PAGE, payload: { page } });

export const changeToNextPage = () => (dispatch, getState) => {
  const {
    vendorInvoices: {
      pagination: { next },
    },
  } = getState();

  dispatch(setPage(next));
};

export const changeToPreviousPage = () => (dispatch, getState) => {
  const {
    vendorInvoices: {
      pagination: { previous },
    },
  } = getState();

  dispatch(setPage(previous));
};

export const changeListing = listing => ({
  type: CHANGE_LISTING,
  payload: {
    per: parseInt(listing, 10),
  },
});

export const calculateQuote = () => (dispatch, getState) => {
  const {
    vendorInvoices: {
      invoices,
      selectedInvoices,
      hasSelectedAllItems,
      token,
      filters,
      simulationTimeoutId,
    },
  } = getState();

  if (hasSelectedAllItems) return {};

  dispatch({ type: QUOTE_INVOICES });

  if (!hasSelectedAllItems) {
    const faceValue = getSumFaceValue(invoices, selectedInvoices);
    dispatch({ type: UPDATE_QUOTE_FACE_VALUE, payload: { faceValue } });
  }

  const newSimulationTimeoutId = debounce(
    () =>
      axios({
        method: 'post',
        url: INVOICES_QUOTE_URL,
        headers: {
          'X-HTTP-Method-Override': 'GET',
          Authorization: `Bearer ${token}`,
        },
        data: {
          ...filters,
          ids: selectedInvoices,
        },
      }).catch(() => {}),
    simulationTimeoutId,
  );

  return dispatch({
    type: UPDATE_SIMULATION_TIMEOUT_ID,
    payload: { simulationTimeoutId: newSimulationTimeoutId },
  });
};

export const quoteAllInvoices = () => (dispatch, getState) => {
  const {
    vendorInvoices: { token, filters },
  } = getState();

  dispatch({ type: QUOTE_ALL_INVOICES });

  return axios
    .get(INVOICES_QUOTE_ALL_URL, {
      headers: { Authorization: `Bearer ${token}` },
      params: { ...filters },
    })
    .catch(() => {});
};

export const resetInvoices = () => ({ type: CLEAR_INVOICES });

export const changeFilters = (token, filters, isAdmin) => {
  if (filters[NUMBER_FILTER]) {
    dispatchDataLayerEvent('titulosFiltrarNumero', isAdmin);
  }

  if (filters[CNPJ_FILTER]) {
    dispatchDataLayerEvent('titulosFiltrarCNPJ', isAdmin);
  }

  if (filters[EXPIRES_EQ_FILTER] || filters[EXPIRES_GT_FILTER] || filters[EXPIRES_LT_FILTER]) {
    dispatchDataLayerEvent('titulosFiltrarVencimento', isAdmin);
  }

  if (
    !filters[NUMBER_FILTER] &&
    !filters[CNPJ_FILTER] &&
    !filters[EXPIRES_EQ_FILTER] &&
    !filters[EXPIRES_GT_FILTER] &&
    !filters[EXPIRES_LT_FILTER]
  ) {
    dispatchDataLayerEvent('titulosLimparFiltro', isAdmin);
  }

  return {
    type: CHANGE_FILTERS,
    payload: { filters },
  };
};

export const requestSimulationReport = () => (dispatch, getState) => {
  const {
    vendorInvoices: { token, selectedInvoices },
  } = getState();

  dispatch({ type: REQUEST_SIMULATION_REPORT });

  return axios
    .post(
      INVOICES_REQUEST_DOWNLOAD_URL,
      { file_format: 'xlsx', ids: selectedInvoices },
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      },
    )
    .catch(() => {});
};

export const showAnticipationLimitError = ({ show, surplusValue }) => ({
  type: UPDATE_SHOW_LIMIT_ERROR,
  payload: { show, surplusValue },
});

export const anticipateInvoices = () => (dispatch, getState) => {
  const {
    vendorInvoices: {
      token,
      selectedInvoices,
      isAdmin,
      availableLimit,
      quote: { netValue },
      roomId,
    },
  } = getState();

  const hasLimitSet = availableLimit !== null && availableLimit !== undefined;

  const surplusValue = parseFloat(netValue) - availableLimit;
  if (hasLimitSet && surplusValue > 0 && selectedInvoices.length) {
    dispatch(showAnticipationLimitError({ show: true, surplusValue }));
    return null;
  }

  dispatchDataLayerEvent(
    selectedInvoices.length === 0 ? 'titulosAnteciparTodos' : 'titulosSolicitarAdiantamento',
    isAdmin,
  );

  dispatch({ type: ANTICIPATE_INVOICES });

  return axios
    .post(
      INVOICES_REQUEST_ANTICIPATION_URL,
      { ids: selectedInvoices, room_id: roomId },
      { headers: { Authorization: `Bearer ${token}` } },
    )
    .catch(() => {
      dispatch({ type: ANTICIPATION_ERROR });
    });
};

export const updateSorting = (sorting = {}) => ({
  type: VENDOR_INVOICES_UPDATE_SORT,
  payload: { sorting },
});

export const updateAvailableLimit = availableLimit => ({
  type: UPDATE_VENDOR_AVAILABLE_LIMIT,
  payload: { availableLimit },
});
