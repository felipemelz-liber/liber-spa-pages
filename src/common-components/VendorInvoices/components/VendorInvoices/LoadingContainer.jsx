import React from 'react';
import PropTypes from 'prop-types';
import { Loading, Bunny } from './VendorInvoices.styles';

const LoadingContainer = ({ loading, children }) => (
  <Loading loading={loading}>
    {loading ? <Bunny loop /> : null}
    {children}
  </Loading>
);

LoadingContainer.propTypes = {
  loading: PropTypes.bool,
  children: PropTypes.node,
};

LoadingContainer.defaultProps = {
  loading: false,
  children: <div />,
};

export default LoadingContainer;
