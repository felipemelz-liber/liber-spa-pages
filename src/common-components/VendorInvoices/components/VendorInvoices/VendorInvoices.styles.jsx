import styled, { createGlobalStyle } from 'styled-components';
import { handleThemeFromObject as getTheme, LoadingBunny, PrimaryLink } from 'liber-components';

export const RootContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`;

export const DisableBootstrap = createGlobalStyle`
  label {
    display: unset;
    margin-bottom: unset;
  }
  th {
    box-sizing: content-box;
  }
`;

export const Title = styled.span`
  font-family: Roboto;
  font-size: 32px;
  font-weight: bold;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.41;
  letter-spacing: 0.13px;
  text-align: left;
  color: ${({ theme }) => getTheme(theme, 'colors.primary', '#009dff')};
`;

export const TitleContainer = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin: 32px 0px;
`;

export const ButtonsContainer = styled.div`
  display: flex;
  align-items: center;
`;

export const HeaderText = styled.span`
  font-family: Roboto;
  font-size: 16px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.38;
  letter-spacing: 0.06px;
  text-align: left;
  color: #6287a7;
  margin-bottom: 56px;
`;

export const TotalInvoices = styled(HeaderText)`
  font-weight: bold;
`;

export const Link = styled(HeaderText).attrs(() => ({ as: 'a' }))`
  font-weight: 500;
  color: ${({ theme }) => getTheme(theme, 'colors.primary', '#009dff')};
  margin-top: 12px;
  margin-bottom: 32px;
  width: fit-content;
`;

export const ButtonLink = styled(PrimaryLink).attrs(() => ({
  version: 2,
  size: 'large',
  outlined: true,
}))`
  margin-right: 16px;
`;

export const Bunny = styled(LoadingBunny)`
  width: 200px;
  height: 250px;
  position: absolute;
  z-index: 202;
`;

export const Loading = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  pointer-events: ${({ loading }) => (loading ? 'none' : null)};
  opacity: ${({ loading }) => (loading ? '0.3' : '1')};
  background-color: white;
  transition: opacity 150ms ease-in-out;
  z-index: 201;
`;

export const EmptyStateContainer = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-end;
  width: 100%;
`;
