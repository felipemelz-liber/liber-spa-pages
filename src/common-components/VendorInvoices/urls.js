const domain = 'http://api.liber.local:3002'; // eslint-disable-line
export const wsDomain = WS_DOMAIN; // eslint-disable-line

export const INVOICES_URL = `${domain}/api/v1/fornecedor/titulos`;
export const INVOICES_QUOTE_ALL_URL = `${domain}/api/v1/fornecedor/titulos/quote_all`;
export const INVOICES_QUOTE_URL = `${domain}/api/v1/fornecedor/titulos/quote`;
export const INVOICES_BANK_ACCOUNTS_URL = `${domain}/api/v1/fornecedor/contas_bancarias`;
export const INVOICES_REQUEST_ANTICIPATION_URL = `${domain}/api/v1/fornecedor/titulos/anticipate_invoices`;
export const INVOICES_REQUEST_DOWNLOAD_URL = `${domain}/api/v1/fornecedor/reports/invoices_simulation`;
export const INVOICES_NEGOTIATIONS_URL = `${domain}/api/v1/fornecedor/titulos/negotiations`;
export const INVOICES_ACTION_CABLE_URL = `${wsDomain}/cable`;
export const INVOICES_ACTION_CABLE_CHANNEL = 'FornecedorScheduleBatchInvoicesChannel';
export const INVOICES_ACTION_CABLE_ROOM = 'fornecedor_schedule_batch_invoices_channel_';
export const ADMIN_INVOICES_URL = `${domain}/api/v1/admin/vendors/:id/invoices`;
export const VENDOR_ANTICIPATION_LIMITS_URL = `${domain}/api/v1/fornecedor/anticipation_limits`;
