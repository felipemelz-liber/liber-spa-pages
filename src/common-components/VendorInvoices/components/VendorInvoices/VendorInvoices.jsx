import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { PrimaryButton, ToastContainer } from 'liber-components';
import { toast } from 'liber-components/components/Toast';
import {
  TitleContainer,
  Title,
  HeaderText,
  TotalInvoices,
  Link,
  RootContainer,
  DisableBootstrap,
  ButtonsContainer,
  ButtonLink,
  EmptyStateContainer,
} from './VendorInvoices.styles';
import {
  fetchInvoices,
  EXPIRES_EQ_FILTER,
  EXPIRES_GT_FILTER,
  NUMBER_FILTER,
  CNPJ_FILTER,
  EXPIRES_LT_FILTER,
  subscribeInvoices,
  anticipateInvoices,
  showAnticipationLimitError,
  subscribeToLimits,
} from '../../actions/VendorInvoicesActions';
import InvoicesTable from '../InvoicesTable/InvoicesTable';
import emptyImage from '../../../../../../assets/images/empty-states/invoices_empty.svg';
import LoadingButton from '../../../Buttons/LoadingButton';
import LoadingContainer from './LoadingContainer';
import AnticipationLimitCard from '../AnticipationLimitCard';
import UnavailableLimitDialog from '../../../UnavailableLimitDialog';

const hasInvoices = (invoices, isFiltered) => invoices.length > 0 || isFiltered;

const getIntroductionText = (hasAvailableInvoices, totalInvoices) => {
  if (hasAvailableInvoices) {
    return (
      <>
        No momento, <TotalInvoices>{totalInvoices}</TotalInvoices> títulos estão disponíveis para
        antecipação. <br />
        Selecione os títulos abaixo e solicite sua antecipação.
      </>
    );
  }

  return (
    <>
      Não encontramos nenhum título disponível para antecipação. <br /> Verifique em{' '}
      <Link href="/fornecedor/negociacoes?q%5Bstatus_in%5D%5B%5D=0&q%5Bstatus_in%5D%5B%5D=7">
        Antecipações
      </Link>{' '}
      se existe alguma operação não concluída.
    </>
  );
};

export const VendorInvoices = ({
  totalInvoices,
  currentPage,
  getInvoices,
  filters,
  loading,
  canInputXml,
  inputXmlPath,
  subscribe,
  pageListing,
  isFiltered,
  invoices,
  selectedInvoices,
  report,
  requestAnticipateAll,
  loadingButton,
  redirectUrl,
  isAdmin,
  sorting,
  token,
  limitError,
  updateShowAnticipationLimitError,
  onSubscribeToLimits,
}) => {
  React.useEffect(() => {
    getInvoices();
  }, [currentPage, pageListing, filters, sorting]);

  if (!isAdmin) {
    React.useEffect(() => {
      subscribe();
      onSubscribeToLimits();
    }, []);

    React.useEffect(() => {
      if (redirectUrl) {
        window.location.href = redirectUrl;
      }
    }, [redirectUrl]);

    React.useEffect(() => {
      const { successful, url, errorMessage } = report;
      if (successful) {
        window.open(url);
      } else if (errorMessage) {
        toast(
          {
            message: 'Erro ao baixar o Relatório da Simulação',
            info: `Ocorreu um erro ao baixar o relatório, tente novamente.
            Se o problema persistir, entre em contato com o suporte`,
          },
          'error',
          8000,
        );
      }
    }, [report]);
  }

  const hasAvailableInvoices = hasInvoices(invoices, isFiltered);

  const renderInvoiceTable = () =>
    hasAvailableInvoices ? (
      <InvoicesTable />
    ) : (
      <EmptyStateContainer>
        <img src={emptyImage} alt="" />
      </EmptyStateContainer>
    );

  const handleCloseLimitDialog = () =>
    updateShowAnticipationLimitError({ show: false, surplusValue: 0 });

  return isAdmin ? (
    <LoadingContainer loading={loading}>
      <div>{renderInvoiceTable()}</div>
    </LoadingContainer>
  ) : (
    <LoadingContainer loading={loading}>
      <ToastContainer />
      <RootContainer>
        <DisableBootstrap />
        <UnavailableLimitDialog
          show={limitError?.show}
          onLeaved={handleCloseLimitDialog}
          surplusValue={limitError?.surplusValue}
        />
        <TitleContainer>
          <Title>Títulos</Title>
          {hasAvailableInvoices ? (
            <ButtonsContainer>
              {canInputXml ? (
                <ButtonLink href={inputXmlPath}>INSERIR NOTAS FISCAIS</ButtonLink>
              ) : null}
              <LoadingButton
                disabled={isFiltered || selectedInvoices.length > 0}
                size="large"
                loading={loadingButton}
                ButtonComponent={PrimaryButton}
                onClick={requestAnticipateAll}
              >
                ANTECIPAR TODOS
              </LoadingButton>
            </ButtonsContainer>
          ) : null}
        </TitleContainer>
        <HeaderText>{getIntroductionText(hasAvailableInvoices, totalInvoices)}</HeaderText>
        <AnticipationLimitCard token={token} />
        {renderInvoiceTable()}
      </RootContainer>
    </LoadingContainer>
  );
};

VendorInvoices.propTypes = {
  totalInvoices: PropTypes.number,
  currentPage: PropTypes.number,
  getInvoices: PropTypes.func,
  filters: PropTypes.shape({
    [NUMBER_FILTER]: PropTypes.arrayOf(PropTypes.number),
    [CNPJ_FILTER]: PropTypes.string,
    [EXPIRES_EQ_FILTER]: PropTypes.string,
    [EXPIRES_GT_FILTER]: PropTypes.string,
    [EXPIRES_LT_FILTER]: PropTypes.string,
  }),
  loading: PropTypes.bool,
  canInputXml: PropTypes.bool,
  inputXmlPath: PropTypes.string,
  subscribe: PropTypes.func,
  pageListing: PropTypes.number,
  isFiltered: PropTypes.bool,
  invoices: PropTypes.arrayOf(PropTypes.object),
  selectedInvoices: PropTypes.arrayOf(PropTypes.number),
  report: PropTypes.shape({
    successful: PropTypes.bool,
    url: PropTypes.string,
    errorMessage: PropTypes.string,
  }),
  requestAnticipateAll: PropTypes.func,
  loadingButton: PropTypes.bool,
  redirectUrl: PropTypes.string,
  isAdmin: PropTypes.bool,
  sorting: PropTypes.shape({}),
  token: PropTypes.string,
  limitError: PropTypes.shape({
    show: PropTypes.bool,
    surplusValue: PropTypes.number,
  }),
  updateShowAnticipationLimitError: PropTypes.func,
  onSubscribeToLimits: PropTypes.func,
};

VendorInvoices.defaultProps = {
  totalInvoices: 0,
  currentPage: 1,
  getInvoices: () => null,
  filters: {
    [NUMBER_FILTER]: null,
    [CNPJ_FILTER]: null,
    [EXPIRES_EQ_FILTER]: null,
    [EXPIRES_GT_FILTER]: null,
    [EXPIRES_LT_FILTER]: null,
  },
  loading: false,
  canInputXml: false,
  inputXmlPath: null,
  subscribe: () => {},
  pageListing: 10,
  isFiltered: false,
  invoices: [],
  selectedInvoices: [],
  report: {},
  requestAnticipateAll: () => {},
  loadingButton: false,
  redirectUrl: null,
  isAdmin: false,
  sorting: {},
  token: '',
  limitError: { show: false, surplusValue: 0 },
  updateShowAnticipationLimitError: () => {},
  onSubscribeToLimits: () => {},
};

const mapStateToProps = ({
  vendorInvoices: {
    pagination: { count: totalInvoices, current: currentPage, per: pageListing },
    filters,
    loading,
    canInputXml,
    inputXmlPath,
    isFiltered,
    invoices,
    selectedInvoices,
    report,
    loadingButton,
    redirectUrl,
    isAdmin,
    sorting,
    token,
    limitError,
  },
}) => ({
  totalInvoices,
  currentPage,
  filters,
  loading,
  canInputXml,
  inputXmlPath,
  pageListing,
  isFiltered,
  invoices,
  selectedInvoices,
  report,
  loadingButton,
  redirectUrl,
  isAdmin,
  sorting,
  token,
  limitError,
});

const mapDispatchToProps = {
  getInvoices: fetchInvoices,
  subscribe: subscribeInvoices,
  requestAnticipateAll: anticipateInvoices,
  updateShowAnticipationLimitError: showAnticipationLimitError,
  onSubscribeToLimits: subscribeToLimits,
};

export default connect(mapStateToProps, mapDispatchToProps)(VendorInvoices);
