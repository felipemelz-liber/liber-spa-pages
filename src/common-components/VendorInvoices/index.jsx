import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import hydrateStore from '../../views/store';
import VendorInvoices from './components/VendorInvoices/VendorInvoices';

const VendorInvoicesComponent = ({ token, vendorId, canInputXml, inputXmlPath, isAdmin }) => (
  <>
    <Provider
      store={hydrateStore({
        vendorInvoices: { token, vendorId, canInputXml, inputXmlPath, isAdmin },
        invoiceDialog: { isAdmin },
      })}
    >
      <VendorInvoices />
    </Provider>
  </>
);

VendorInvoicesComponent.propTypes = {
  token: PropTypes.string.isRequired,
  vendorId: PropTypes.number.isRequired,
  inputXmlPath: PropTypes.string,
  canInputXml: PropTypes.bool,
  isAdmin: PropTypes.bool,
};

VendorInvoicesComponent.defaultProps = {
  inputXmlPath: '',
  canInputXml: false,
  isAdmin: false,
};

export default VendorInvoicesComponent;
