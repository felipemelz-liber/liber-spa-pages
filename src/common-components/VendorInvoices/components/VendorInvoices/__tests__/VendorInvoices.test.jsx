import React from 'react';
import sinon from 'sinon';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import * as toastObj from 'liber-components/components/Toast/actions';
import { VendorInvoices } from '../VendorInvoices';

describe('VendorInvoices tests', () => {
  configure({ adapter: new Adapter() });

  const getInvoices = sinon.spy();
  const subscribe = sinon.spy();
  const toastSpy = sinon.spy(toastObj, 'toast');
  sinon.stub(React, 'useEffect').callsArg(0);
  const windowSpy = sinon.spy();
  global.open = windowSpy;

  afterEach(() => {
    sinon.resetHistory();
  });

  afterAll(() => {
    sinon.restore();
  });

  it('should render correctly without props', () => {
    const wrapper = shallow(<VendorInvoices />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly with loading', () => {
    const wrapper = shallow(<VendorInvoices loading />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly whend vendor can input xml', () => {
    const wrapper = shallow(<VendorInvoices canInputXml />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call subscribe and getInvoices on render', () => {
    shallow(<VendorInvoices getInvoices={getInvoices} subscribe={subscribe} />);
    expect(getInvoices.called).toBeTruthy();
    expect(subscribe.called).toBeTruthy();
  });

  it('should call window open when report successfuly generated', () => {
    const report = { successful: true, url: 'mock' };
    shallow(<VendorInvoices report={report} />);

    expect(windowSpy.calledWith(report.url)).toBeTruthy();
  });

  it('should call toast when report is unsuccessful', () => {
    const report = { successful: false, errorMessage: 'error' };
    shallow(<VendorInvoices report={report} />);

    expect(toastSpy.called).toBeTruthy();
  });
});
