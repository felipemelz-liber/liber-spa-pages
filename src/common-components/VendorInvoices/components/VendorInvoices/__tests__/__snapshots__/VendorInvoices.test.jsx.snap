// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VendorInvoices tests should render correctly whend vendor can input xml 1`] = `
<LoadingContainer
  loading={false}
>
  <_default />
  <styled.div>
    <GlobalStyleComponent />
    <UnavailableLimitDialog
      onLeaved={[Function]}
      redirectUrl={null}
      show={false}
      surplusValue={0}
    />
    <styled.div>
      <styled.span>
        Títulos
      </styled.span>
    </styled.div>
    <styled.span>
      Não encontramos nenhum título disponível para antecipação. 
      <br />
       Verifique em
       
      <Styled(styled.span)
        href="/fornecedor/negociacoes?q%5Bstatus_in%5D%5B%5D=0&q%5Bstatus_in%5D%5B%5D=7"
      >
        Antecipações
      </Styled(styled.span)>
       
      se existe alguma operação não concluída.
    </styled.span>
    <AnticipationLimitCard
      token=""
    />
    <styled.div>
      <img
        alt=""
        src="invoices_empty.svg"
      />
    </styled.div>
  </styled.div>
</LoadingContainer>
`;

exports[`VendorInvoices tests should render correctly with loading 1`] = `
<LoadingContainer
  loading={true}
>
  <_default />
  <styled.div>
    <GlobalStyleComponent />
    <UnavailableLimitDialog
      onLeaved={[Function]}
      redirectUrl={null}
      show={false}
      surplusValue={0}
    />
    <styled.div>
      <styled.span>
        Títulos
      </styled.span>
    </styled.div>
    <styled.span>
      Não encontramos nenhum título disponível para antecipação. 
      <br />
       Verifique em
       
      <Styled(styled.span)
        href="/fornecedor/negociacoes?q%5Bstatus_in%5D%5B%5D=0&q%5Bstatus_in%5D%5B%5D=7"
      >
        Antecipações
      </Styled(styled.span)>
       
      se existe alguma operação não concluída.
    </styled.span>
    <AnticipationLimitCard
      token=""
    />
    <styled.div>
      <img
        alt=""
        src="invoices_empty.svg"
      />
    </styled.div>
  </styled.div>
</LoadingContainer>
`;

exports[`VendorInvoices tests should render correctly without props 1`] = `
<LoadingContainer
  loading={false}
>
  <_default />
  <styled.div>
    <GlobalStyleComponent />
    <UnavailableLimitDialog
      onLeaved={[Function]}
      redirectUrl={null}
      show={false}
      surplusValue={0}
    />
    <styled.div>
      <styled.span>
        Títulos
      </styled.span>
    </styled.div>
    <styled.span>
      Não encontramos nenhum título disponível para antecipação. 
      <br />
       Verifique em
       
      <Styled(styled.span)
        href="/fornecedor/negociacoes?q%5Bstatus_in%5D%5B%5D=0&q%5Bstatus_in%5D%5B%5D=7"
      >
        Antecipações
      </Styled(styled.span)>
       
      se existe alguma operação não concluída.
    </styled.span>
    <AnticipationLimitCard
      token=""
    />
    <styled.div>
      <img
        alt=""
        src="invoices_empty.svg"
      />
    </styled.div>
  </styled.div>
</LoadingContainer>
`;
